src/pds/cross_report.rs

```diff
@@ -0,0 +1,404 @@
+        let mut unfiltered_report =
+            self.request.compute_report(&relevant_events);
+
+        // only keep the buckets that are requested
+        unfiltered_report.bin_values.retain(|bucket, _| {
+            relevant_event_selector.requested_buckets.contains(bucket)
+        });
+
+        let mut oob_filters = vec![];
+        for epoch_id in epochs {
+            let epoch_relevant_events = self.events.for_epoch(&epoch_id);
+
+            // Step 2. Compute individual loss for current epoch.
+            let individual_privacy_loss = compute_epoch_loss(
+                &self.request,
+                epoch_relevant_events,
+                &unfiltered_report,
+                num_epochs,
+            );
+
+            let filter_id =
+                FilterId::PerQuerier(epoch_id, beneficiary_uri.clone());
+            let filter_status = filter_storage
+                .try_consume(&filter_id, &individual_privacy_loss)?;
+
+            if filter_status == FilterStatus::OutOfBudget {
+                // Not enough budget, drop events without any filter
+                // consumption
+                relevant_events.drop_epoch(&epoch_id);
+
+                // Keep track of why we dropped this epoch
+                oob_filters.push(filter_id);
+            }
+        }
+
+        // Now that we've dropped OOB epochs, we can compute the final report.
+        let mut filtered_report = self.request.compute_report(&relevant_events);
+
+        // only keep the buckets that are requested
+        filtered_report.bin_values.retain(|bucket, _| {
+            relevant_event_selector.requested_buckets.contains(bucket)
+        });
+
+        let report = PdsReport {
+            filtered_report,
+            unfiltered_report,
+            oob_filters,
+        };
+        Ok(report)
+    }
+}
+
+impl<U, FS, ERR> PrivateDataServiceCore<PpaHistogramRequest<U>, FS, ERR>
+where
+    U: Uri,
+    FS: FilterStorage<
+        FilterId = FilterId<PpaEpochId, U>,
+        Budget = PureDPBudget,
+    >,
+    ERR: From<FS::Error>,
+{
+    pub fn measure_conversion(
+        &mut self,
+        request: PpaHistogramRequest<U>,
+        mut relevant_events: RelevantEvents<PpaEvent<U>>,
+    ) -> Result<AttributionObject<PpaHistogramRequest<U>>, ERR> {
+        let uris = request.report_uris();
+
+        let epochs = request.epoch_ids();
+        let num_epochs = epochs.len();
+
+        // Compute the raw report, useful for debugging and accounting.
+        let unfiltered_result = request.compute_report(&relevant_events);
+
+        let mut oob_filters = vec![];
+        for epoch_id in epochs {
+            let individual_privacy_loss =
+                request.histogram_multi_epoch_report_global_sensitivity();
+
+            let source_losses = compute_epoch_source_losses(
+                &request,
+                relevant_events.sources_for_epoch(&epoch_id),
+                &unfiltered_result,
+                num_epochs,
+            );
```

Comment:
In the pseudocode I'm using 2a^max for source losses too (eps^i), instead of the tighter individual budget.

Reply:
Done in https://github.com/columbia/pdslib/pull/95/commits/925d709a57e6f5957bbe9b603604b1c2c1fa53e3